containers.jl

```diff
@@ -176,6 +178,24 @@ addconstraint!(c::ErasedContainer, fs::Tuple) = push!(c.constraints, fs)
 
 eachconstraint(f, c::ErasedContainer) = foreach(f, c.constraints)
 
+"""
+Uses a TypeSortedCollection to store the constraints.
+"""
+mutable struct TypeSortedContainer
+    constraints::TypeSortedCollection
+    TypeSortedContainer() = new(TypeSortedCollection{Tuple{}}())
+end
+
+function addconstraint!(c::TypeSortedContainer, fs::Tuple{F, S}) where {F <: AbstractFunction, S <: AbstractSet}
+    if fs isa eltype(c.constraints)
```

Comment:
What about putting a function barrier around this? That is, could you pass `c.constraints` to an inner function that would do this check in a type-stable manner?

Reply:
At first I thought no, because in one case you actually need to set a field of `c`, but I found a way that's a bit faster. I'll update the benchmark results.